### PR TITLE
Implement JellySquid's Entity Collision optimisations patch

### PR DESCRIPTION
**Pull request by smonnnn**
This patch replaces the vanilla method for block collisions with a much faster one, which uses simple collision testing against full-cube block shapes and faster world border checks which do not go through the voxelsystem. It also replaces the method for getting entity collisions with a version that makes less allocations and without stream code. 

The original code is written by JellySquid, and can be found [here](https://github.com/jellysquid3/lithium-fabric/blob/1.15.x/fabric/src/main/java/me/jellysquid/mods/lithium/common/shapes/LithiumEntityCollisions.java)
The code is licensed under GNU Lesser General Public License v3.0.

I have implemented it in the same way as it was originally implemented in Lithium, I believe this will make it easier to update the patch in the future since you can then take the original updated yarn mapped counterpart from Lithium and remap that to spigot mappings.

### CODE REVIEW

File: Spigot-Server-Patches/0491-Implement-JellySquid-s-Entity-Collision-optimisation.patch [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
++                    int y = cuboidIt.getY();
++                    int z = cuboidIt.getZ();
++
++                    int edgesHit = cuboidIt.getEdgeCoordinatesCount();
++
++                    if (edgesHit == 3) {
++                        continue;
++                    }
++
++                    IBlockAccess chunk = world.c(x >> 4, z >> 4); // c() = getExistingChunk()
```
Comment (aikar) [inline Spigot-Server-Patches/0491-Implement-JellySquid-s-Entity-Collision-optimisation.patch:76]: consider https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0454-Optimize-Collision-Chunk-lookup-and-avoid-loading-fa.patch
here

I didn't know enough about this system, but if the chunk is null, we really should be returning a full block collision to stop the movement entirely into the unloaded chunk.

Returning null here would consider it as a non collision, can we fix it to mark it as a collision?

File: Spigot-Server-Patches/0491-Implement-JellySquid-s-Entity-Collision-optimisation.patch [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+             throw (IllegalArgumentException) SystemUtils.c(new IllegalArgumentException());
+diff --git a/src/main/java/net/minecraft/server/WorldBorder.java b/src/main/java/net/minecraft/server/WorldBorder.java
+index 5f10c4338..81636c2ba 100644
+--- a/src/main/java/net/minecraft/server/WorldBorder.java
++++ b/src/main/java/net/minecraft/server/WorldBorder.java
+@@ -48,6 +48,8 @@ public class WorldBorder {
+         return this.b(entity.locX(), entity.locZ());
+     }
+ 
++    public final VoxelShape asVoxelShape(){ return a();} //Paper - OBFHELPER
```
Comment (aikar) [inline Spigot-Server-Patches/0491-Implement-JellySquid-s-Entity-Collision-optimisation.patch:368]: need to move this down 1 line so it borders.
that single line can be enough to throw off patch apply conflicts (which we want!)

**Review by aikar — CHANGES_REQUESTED** [review]
Overall looks great! Only 1 real feedback id love if we can fix in this about  chunk unloaded state.

See linked patch, if we can change it to getChunkAtIfLoaded and improve null behavior (to not even need the distance checks I had), it would be great.

**Comment by smonnnn** [comment]
Having some issues with it not wanting to patch out of the blue, I'm going to fix it and pr again later.

**Comment by aikar** [comment]
Manual merged with https://github.com/PaperMC/Paper/commit/d454bbd5e198d0c6d4c773f55427a92197fbc435

**Pull request closed without merging**